notebook/_posts/2021-01-30-pico8-offset.md

```diff
@@ -0,0 +1,294 @@
+
+Awesome! With the functions above, we now have a rudimentary system for
+movement. This works, but we just need to draw the sprite! In the `_draw()`
+function, just call the [`spr()` function](https://pico-8.fandom.com/wiki/Spr):
+
+```lua
+-- draw_game is a handler for all drawing events
+function draw_game()
+    cls(0)
+    map()
+    spr(1, p.x, p.y, 1, 1, p.flip)
+end
+```
+
+<!-- TODO: insert pico8-js of current output -->
+
+## Smoother tile transition
+
+Picollino's movement still looks a bit jarring: he seems to "teleport" from one
+tile to another. This is because we are moving 8 pixels right away, a large
+distance in my opinion. I'll improve animation by doing the following:
+1. Flip the sprite when moving to the opposite direction
+2. Add a "smoother" to ease the transition from one tile to the next.
+3. Sprite animation
+
+### Flip sprite based on direction
+
+The first one is easy. Recall that in our `make_player()` function, we have
+a `flip` attribute. We set that as `true` when our sprite is moving to the left,
+i.e. when `dx < 0`. We'll update the `move_player()` function to account for
+this change.
+
+```lua
+-- move_player updates player coordinates
+-- @param dx change in the x-axis 
+-- @param dy change in the y-axis
+function move_player(dx, dy)
+    local destx, desty = p.x + dx * 8, p.y + dy * 8
+    ------------------NEW CODE---------------------
+    if dx < 0 then
+        p.flip = true
+    elseif dx > 0 then 
+        p.flip = false
+    end
+    -----------------------------------------------
+    p.x, p.y = destx, desty
+end
+```
+
+We use `elseif` instead of `else` so that the sprite won't immediately
+face `right` if our last action was `left` and we decide to move `up` or
+`down`.
+
+### Tile transition
+
+In order to understand LazyDev's technique for smoother tile transition,  let
+me  illustrate an example. Suppose we want to move a sprite to the right from
+tile $$T(3,2)$$ to $$(4,2)$$. In pixel coordinates, we are essentially
+"teleporting" from $$P(24,16)$$ to $$P(32,16)$$, a huge distance that doesn't
+look like walking. Instead, we want to traverse the $$x$$ axis in small
+increments, i.e. $$24, 24.8, 25.6, 26.4, \ldots$$, until it reaches its
+destination.
+
+<!-- TODO: insert gif -->
+
+We accomplish this by defining an offset that will incrementally move the
+sprite from its old coordinate to the new one, i.e. $$P_{x_2,y_2} = P_{x_1,y_1} + o_{x,y}$$. The rate in which the offset updates is controlled by a new
+variable, $$p_t$$, that dictates how fast the transition will be animated.
+
+Because we want to save tokens, we precompute the destination and subtract the
+offset as we go along. So in our earlier example, we will set $$P_{x,y}$$ to
+its new value, $$(32,16)$$, and perform subtraction when calling the `spr()`
+function. Because we started from $$(24,16)$$, we subtract $$8$$ from $$32$$,
+then $$8-o_x$$, and so on and so forth until $$o_x=0$$.
```

Comment:
Explain this a bit further, seemed unclear:
* You subtract 8 from 32, 32-8
* Then you subtract 7.X from 32, 32 - (8-o)

Maybe we should rewrite it into 8+o_x?
o_x should decrease until it reaches zero